A55-xds-stateful-session-affinity.md

```diff
@@ -241,47 +256,49 @@ Let's call the new policy `xds_override_host_experimental`. This policy
 contains subchannel management, endpoint management, and RPC routing logic as
 follows:
 
-* maintain a map - let’s call this `overrideHostMap`. The key is (IP:port)
+* maintain a map - let's call it `address_map`. The key is (IP:port)
   and the value is the tuple
-  `[subchannel, list of other equivalent addresses]`. When a host (endpoint)
-  has multiple addresses (e.g. IPv6 vs IPv4 or due to multiple network
-  interfaces) they are said to be equivalent addresses. Whenever we get an
-  updated list of addresses from the resolver, we create an entry in the map
-  for each address, and populate the list of equivalent addresses. The
-  subchannel is based on subchannel creation/destruction or
-  connection/disconnection from the child policy. The equivalent address list
-  computation on each resolver update is performed as per the following
-  pseudo-code:
+  `[subchannel, EDS health-status, list of other equivalent addresses]`.
+  When a host (endpoint) has multiple addresses (e.g. IPv6 vs IPv4 or due to
+  multiple network interfaces) they are said to be equivalent addresses.
+  Whenever we get an updated list of addresses from the resolver, we create an
+  entry in the map for each address, and populate the list of equivalent
+  addresses and EDS health-status. The subchannel is based on subchannel
+  creation/destruction or connection/disconnection from the child policy. The
+  equivalent address list computation on each resolver update is performed as
+  shown in the following pseudo-code:
 
 ```
 // Find the transitive closure of all EAGs.
-eags_to_process = [set(eag) for eag in resolver_update]
-completed_eag_sets = []
+eags_to_process = [set([eag,health_status]) for eag in resolver_update]
```

Comment:
I think the latest changes here made this pseudo-code much harder to follow, and it's not actually consistently correct -- e.g., the "if eag.intersects(current_eag)" line does not account for the health status.

I think we need to take a small step back here and answer the question of how we would handle the case where a given address exists in multiple EAGs but with different health status values.  That's the only case that really causes problems for this algorithm, because if all instances of a given address always have the same health status, then we don't need to consider health status in this algorithm at all.

One option I can see would be to consider each unique (address, health_status) tuple as a separate address (i.e., address A with health status `HEALTHY` and the same address A with health status `DRAINING` would be considered distinct entries).  In that case, we don't need any changes to the pseudo-code here; we can just use the previous version of the pseudo-code with a comment that each entry in an EAG is an (address, health_status) tuple.

Another option would be to say that we basically always expect all occurrances of a given address to have the same health status.  In that case, we would still use the previous version of the pseudo-code as-is, and we can set the health status value of the address based on any of its occurances in the EAG list.

Stepping back a bit further, another thing we may want to consider here is how Envoy would react in the case where there are multiple hosts with the same address.  Which one would it use?  @wbpcode, can you tell us what the behavior would be in that case?  I'm wondering if maybe our algorithm here is too complex and maybe we can do something simpler.

Reply:
@markdroth 

At a minimum we need to record the health-status in each entry of `lb_policy.address_map` because the picker code needs it, so that was needed to be added.

About your point about how to handle conflicting health-statuses, I don't have strong opinions one way or the other. The one thing I would like to say is EAG currently only exists in Java and there too we don't get any real EAGs (having more than one address) from the xDS resolver so this is really an academic point at this point. So I agree with you about using the old pseudo-code as-is with only the required change to add health-status. I believe the current pseudo-code basically reflects your suggestion in paragraph 4 of your comment - do you agree?

Reply:
I have restored the pseudo-code to what it was before the commit 70a36d591005aaf816943dcacab5ef7f9a819729 . However we still need to indicate that the health_status is to be recorded somewhere in the `address_map`

Reply:
@wbpcode @htuch @yanavlasov I'd really like input on this issue from the Envoy side.

What does Envoy do if the same address appears multiple times in the cluster?  Does it just de-dup them?  What if the two instances of the same address are in different priorities or different localities?

I see in the Envoy code that the LB policies use a host map that is keyed by the "ip:port" string of the host.  Does this mean that it's just impossible for there to be two instances of the same address in the same cluster?  If so, then this isn't a conflict for stateful session affinity, but then I'd like to understand how the dups are handled more generally.

Reply:
As per https://github.com/envoyproxy/envoy/issues/23445, it looks like Envoy does not allow duplicate addresses within a given cluster.  So I think we'll do the same thing in gRPC, which means that we don't need to worry about duplicates here.

I think this means that we don't need any of the complexity here for handling duplicates.  Instead, I think we can do this:

```
# Resolver update contains a list of EquivalentAddressGroups,
# each of which has a list of addresses and a health status.
all_addresses = set()
for eag in resolver_update:
  for address in eag.addresses:
    lb_policy.address_map[address].equivalent_addresses = eag.addresses
    lb_policy.address_map[address].health_status = eag.health_status
    all_addresses.insert(address)
# Now remove equivalencies for any address not found in the current update.
for address, entry in lb_policy_map.address_map:
  if address not in all_addresses:
    entry.equivalent_addresses.clear()
```

Reply:
changed